src/libtomahawk/audio/AudioEngine_p.h

```diff
@@ -1,3 +1,4 @@
+#include "AudioEngine.h"
```

Comment:
A forward declaration should be enough.

Reply:
it should be enough, but it causes compilation errors in llvm/clang for me.

Reply:
What version of clang? What errors?

Reply:
one moment, ill copy them in

Reply:
```
In file included from src/libtomahawk/51e5690@@tomahawk@sha/moc_AudioEngine_p.cpp:10:
src/libtomahawk/51e5690@@tomahawk@sha/../../../../src/libtomahawk/audio/AudioEngine_p.h:16:36: error: expected ‘)’ before ‘*’ token
   16 |     AudioEnginePrivate( AudioEngine* q )
      |                       ~            ^
      |                                    )
src/libtomahawk/51e5690@@tomahawk@sha/../../../../src/libtomahawk/audio/AudioEngine_p.h:23:5: error: ‘AudioEngine’ does not name a type
   23 |     AudioEngine* q_ptr;
      |     ^~~~~~~~~~~
In file included from ../src/libtomahawk/DllMacro.h:23,
                 from src/libtomahawk/51e5690@@tomahawk@sha/../../../../src/libtomahawk/audio/AudioOutput.h:25,
                 from src/libtomahawk/51e5690@@tomahawk@sha/../../../../src/libtomahawk/audio/AudioEngine_p.h:2,
                 from src/libtomahawk/51e5690@@tomahawk@sha/moc_AudioEngine_p.cpp:10:
src/libtomahawk/51e5690@@tomahawk@sha/../../../../src/libtomahawk/audio/AudioEngine_p.h:24:24: error: ‘AudioEngine’ does not name a type
   24 |     Q_DECLARE_PUBLIC ( AudioEngine )
      |                        ^~~~~~~~~~~
/usr/include/qt/QtCore/qglobal.h:1152:12: note: in definition of macro ‘Q_DECLARE_PUBLIC’
 1152 |     inline Class* q_func() { return static_cast<Class *>(q_ptr); } \
      |            ^~~~~
src/libtomahawk/51e5690@@tomahawk@sha/../../../../src/libtomahawk/audio/AudioEngine_p.h:24:24: error: ‘AudioEngine’ does not name a type
   24 |     Q_DECLARE_PUBLIC ( AudioEngine )
      |                        ^~~~~~~~~~~
/usr/include/qt/QtCore/qglobal.h:1153:18: note: in definition of macro ‘Q_DECLARE_PUBLIC’
 1153 |     inline const Class* q_func() const { return static_cast<const Class *>(q_ptr); } \
      |                  ^~~~~
In file included from src/libtomahawk/51e5690@@tomahawk@sha/moc_AudioEngine_p.cpp:10:
src/libtomahawk/51e5690@@tomahawk@sha/../../../../src/libtomahawk/audio/AudioEngine_p.h:54:12: error: ‘AudioEngine’ does not name a type
   54 |     static AudioEngine* s_instance;
```
@dschmidt

Reply:
And you added `class AudioEngine;` under the includes?

Reply:
i added `#include AudioEngine.h` and that fixed it.

Reply:
try class AudioEngine; instead

https://en.wikipedia.org/wiki/Forward_declaration

Reply:
?? i dont see why thats necessary, when including the missing error works just fine

Reply:
Obviously you don't see the article I just pasted you, let me quote:

```
Forward-declaration is used to avoid unnecessary coupling which help reducing compilation time by reducing the number of header inclusion. This has a triple advantage:

- reduce the number of files opened by #include (hence the number of operating system calls)
- reducing the volume of the pre-processed files (as the header is not included)
- reducing recompilation impact when the forward declared class is modified.
```

Reply:
> ?? i dont see why thats necessary, when including the missing error works just fine

It is fine if you'd like to contribute and have limited c++ experience and knowledge, but please take the advise from the project developers.

Reply:
> Obviously you don't see the article I just pasted you, let me quote:
> 
> ```
> Forward-declaration is used to avoid unnecessary coupling which help reducing compilation time by reducing the number of header inclusion. This has a triple advantage:
> 
> - reduce the number of files opened by #include (hence the number of operating system calls)
> - reducing the volume of the pre-processed files (as the header is not included)
> - reducing recompilation impact when the forward declared class is modified.
> ```

as this header is passed to a preprocessor, and explicit definition is the best option imo

Reply:
Every file in C++ is passed to a preprocessor, what's the point?
What's an explicit definition after all?

Please provide me with the error you get when forward declaring.

Reply:
im not here to argue with you about c++ syntax. you either want the pull requests, or you dont.

Reply:
To make this clear:

We invested a lot of our time into this project and while we're not actively working on it, we still care about it.

So yes, if you want things merged, you are here to argue with us about your changes. If you refuse to do so, you won't see your changes upstream. Simple as that.